Update multinode demo (#245)
timescale/docs.timescale.com-content: tutorials/clustering.md

writes and queries to multiple data nodes. For more information on the actual
underlying architecture, you can view our [architecture][architecture] section.

This tutorial will walk you through setting up a three-node distributed

> database and creatint a distributed hypertable. You will learn how to enable

Reviewer: ```suggestion
database and creating a distributed hypertable. You will learn how to enable
```

Author: fixed

writes and queries to multiple data nodes. For more information on the actual
underlying architecture, you can view our [architecture][architecture] section.

In this tutorial, we will walk through setting up a three node distributed
database and creating a distributed hypertable. As part of this we will show
you how to enable `scram-sha-256` authentication and add new roles to your

> distributed database.

Reviewer 2: Suggestion: "This tutorial will walk you through setting up a three-node distributed database and create a distributed hypertable. You will learn how to enable password authentication and add new roles to your distributed database."

will assume the `postgres` role for this, but you may substitute any
role common to all machines with superuser permission.

### Update your Postgres configuration

First of all, if you haven't yet run the `timescaledb-tune` utility, it's
highly recommended that you do so before starting. Do this for all of your
nodes.
```
sudo timescaledb-tune
```

In addition, make sure you've enabled prepared transactions on your data
nodes. This is essential for allowing Timescale distributed transactions to

> work. prepared transactions are disabled by default, so make sure you have

Reviewer: ```suggestion
work. Prepared transactions are disabled by default, so make sure you have
```

Author: fixed

this line in `postgresql.conf` on every data node.
```
max_prepared_transactions = 150
```

### Set password encryption

You are now ready to set up the network communication between the nodes and

> add authentication. For more details and option for this step, please visit

Reviewer: ```suggestion
add authentication. For more details and options for this step, please visit
```

Author: Fixed

the [data node authentication][data-node-authentication] page.

In this example you'll be adding `scram-sha-256` password authentication to your
nodes. The first step of this is to enable password encryption on all the nodes
by adding the following to `postgresql.conf`:

```
password_encryption = 'scram-sha-256'
```

>:WARNING: This setting is not automatically applied to any existing passwords.
Any existing roles will have to update to new passwords to take advantage of
the encryption. You can do this in Postgres by using the `ALTER ROLE` command

> to specify the new password (this can be the same as the old password).

Reviewer: This is described in better details in the section to set up the data nodes. May be give a link  instead?

Author: The example in that section is actually intended to solve the issue of the postgres role not having a password set up.  I'm not sure it would make sense to the user to point them at that section, though I did add another sentence there explaining how to fix the problem if they run the command too soon.  For this section, I added a link to the postgres ALTER ROLE documentation.

will assume the `postgres` role for this, but you may substitute any
role common to all machines with superuser permission.

### Update your Postgres configuration

First of all, if you haven't yet run the `timescaledb-tune` utility, it's
highly recommended that you do so before starting. Do this for all of your

> nodes.

Reviewer 2: Suggestion for a less formal version: "It is highly recommended that you run `timescaledb-tune` on all nodes before you start, so if you haven't done this yet, do it now."

```
sudo timescaledb-tune
```

> Also, while you're visiting each of your nodes, make sure you've enabled

Reviewer: And due to mix of `your` and `our`, `the` can be used instead:
```suggestion
Also, while you're visiting each of the data nodes, make sure you've enabled
```

> prepared transactions on them, as these are essential for our distributed

Reviewer 3: Is this strictly necessary for the access node? The PG docs recommends keeping this setting a zero if possible.

transactions. By default these are disabled, so add the following line to

> your `postgresql.conf`.

Reviewer 2: Suggestion: "Make sure you have enabled prepared transactions on all nodes. This is is essential for the distributed transactions to work. Prepared transactions are by default disabled, so make sure that you have this line in your `postgresql.conf`."

> your `postgresql.conf`.

Reviewer: ```suggestion
`postgresql.conf` on every data node.
```

```
max_prepared_transactions = 150
```

### Set password encryption

> The next thing we'll be doing is setting up the network communication between

Reviewer 2: ```suggestion
You are now ready to set up the network communication between
```

> our nodes and adding authentication. For more details and option for this step,

Reviewer 2: ```suggestion
the nodes and add authentication. For more details and option for this step,
```

> our nodes and adding authentication. For more details and option for this step,

Reviewer: I suggest to move the sentence about more details to the end of this section, since the details are in addition to the information in the section. Otherwise it feels insecure if the link can be skipped by the reader.

please visit the [data node authentication][data-node-authentication] page.

In this example we'll be adding `scram-sha-256` password authentication to your
nodes and adding a password entry for the postgres role as well as a new user

> role we'll be creating later on. The first step of this is to enable password

Reviewer 2: Quite a long sentence. Rewrite?

encryption on all the nodes by adding the following to `postgresql.conf`:

```
password_encryption = 'scram-sha-256'
```

>:WARNING: This setting is not automatically applied to any existing passwords.
Any existing roles will have to update to new passwords to take advantage of

> the encryption.

Reviewer 2: How do you do that? Please describe.

After making this change, restart the postgres service so the changes made so
far can take effect.

### Set up the access node

The next step is to enable `access.example.com` to send the passwords for the

> various users of the system. Create a new file `passfile` in the Postgres data

Reviewer 2: ```suggestion
users of the system. Create a new file `passfile` in the Postgres data
```

directory, and add the following lines:
```
*:*:*:postgres:postgres
*:*:*:testuser:testpass
```

This will result in the access node always using password `postgres` for role
`postgres` and `testpass` for role `testuser` whenever connecting to a data

> node. Make sure that this file is readable by the postgres user.

Reviewer 3: Note that there is a difference here between the `postgres` user on the system (OS) and the `postgres` user in the DB. In fact, the former might not exist at all on some OSes (i.e., postmaster runs under another user name). I guess what you want to say is that the whatever user the postmaster runs under needs to be able to read this file.

### Set up the data nodes

> Next we will set up the data node to accept incoming connections from the

Reviewer 3: "data node" or "data nodes"?

> Next we will set up the data node to accept incoming connections from the

Reviewer 2: ```suggestion
The next step is to set up the data node to accept incoming connections from the
```

> access node for the roles we expect. The first thing you need to do here is

Reviewer 2: ```suggestion
access node for the expected roles. The first thing you need to do here is
```

to make sure that Postgres will listen for incoming connections. Add the
following to `postgresql.conf`:
```
listen_addresses = '*'
```

Next, add the roles you want to allow to `pg_hba.conf`:
```
host    all    postgres    access.example.com    scram-sha-256
host    all    testuser    access.example.com    scram-sha-256
```

This allows connections to the `postgres` and `testuser` roles, but only from
the access node, and only using `scram-sha-256` authentication.

> Finally, we need to make sure the `postgres` role's password matches the

Reviewer 2: ```suggestion
Finally, make sure the `postgres` role's password matches the
```

> password we set in the `passfile` on the access node. Run the following

Reviewer 2: ```suggestion
password in the `passfile` on the access node. Run the following
```

> command in psql.

Reviewer 3: make clear that `psql` must connect to the data node. In fact, probably better to say "run the following command on all data nodes."

```sql
ALTER ROLE postgres PASSWORD 'postgres';
```

Note that if you did not restart postgres after changing the
`password_encryption` above, the password will not be properly encrypted and
you will not be able to connect to this data node as `postgres`.

> Restart Postgres on all of your nodes one last time to ensure it is running

Reviewer: English is not my native language, but I feel `one last time` is not correct. I guess it should be either `one more time` or `last time`.

Reviewer: Thank you to clarifying! Now I learned about this phrase. I guess it fits better here.

SELECT * FROM timescaledb_information.data_node;
```

### Create a new role

> Now that we've created a database and added a couple data nodes, let's go ahead

Reviewer 2: ```suggestion
Now that you have created a database and added a couple data nodes, let's go ahead
```

> Now that we've created a database and added a couple data nodes, let's go ahead

Reviewer: It is easy to be explicit:
```suggestion
Now that we've created a database and added two data nodes, let's go ahead
```

and create a new user role we can use for our distributed database. We'll use
Timescale's `distributed_exec` function to perform this action on the data
nodes:
```sql
CREATE ROLE testuser WITH LOGIN PASSWORD 'testpass';

> SELECT * FROM distributed_exec($$ CREATE ROLE testuser WITH LOGIN PASSWORD 'testpass' $$);

Reviewer 2: Even though this is possible, isn't this a security issue? The connection is not secure, is it? No need to remove the command, but might be a good idea to highlight that this can be insecure.

Author: Depends whether postgres is using SSL.  I added a comment to the following paragraph.

```

> This creates the same user role on all the data nodes. It's important that we

Reviewer 2: ```suggestion
This creates the same user role on all the data nodes. It is important that
```

Author: good catch, fixed

> create the user with `LOGIN` permission for them to perform distributed

Reviewer 2: ```suggestion
the user is created with `LOGIN` permission for them to perform distributed
```

Author: changed

SELECT * FROM timescaledb_information.data_node;
```

### Create a new role

Now that you have created a database and added a couple data nodes, let's go
ahead and create a new user role we can use for our distributed database.
We'll use Timescale's `distributed_exec` function to perform this action on

> the data nodes:

Reviewer: Entire tutorial uses `you`, but here it switches to `we`. May be to use `you` here too?

Author: Good catch, I've been trying to fix up these pronouns from the original tutorial.
